cpp/src/io/parquet/page_enc.cu

```diff
@@ -985,15 +974,14 @@ __global__ void __launch_bounds__(128, 8) gpuEncodePages(EncPage *pages,
       while (s->rle_numvals < s->page.num_rows) {
         uint32_t rle_numvals = s->rle_numvals;
         uint32_t nrows       = min(s->page.num_rows - rle_numvals, 128);
-        uint32_t row         = s->page.start_row + rle_numvals + t;
+        uint32_t row         = s->page.start_row + rle_numvals + t + s->col.leaf_column_offset;
         // Definition level encodes validity. Checks the valid map and if it is valid, then sets the
         // def_lvl accordingly and sets it in s->vals which is then given to RleEncode to encode
-        uint32_t def_lvl =
-          (rle_numvals + t < s->page.num_rows && row < s->col.num_rows)
-            ? (valid != nullptr)
-                ? (valid[(row + validity_offset) / 32] >> ((row + validity_offset) % 32)) & 1
-                : 1
-            : 0;
+        uint32_t def_lvl = (rle_numvals + t < s->page.num_rows && row < s->col.num_rows)
+                             ? s->col.leaf_column->is_valid(row)
+                             : 0;
+        // Non-list leaf column does not require taking into account
+        // leaf_column_offset
```

Comment:
I admit this was confusing. I am referring to the fact that on line 982-984 we do not need to add s->col.leaf_column_offset to row because s->col.leaf_column_offset evaluates to 0 for all non list column and this whole if block works on non list column.
I could move this before those lines and number the comments to denote that there are two separate things being talked about here.

Reply:
yeah, it should be before those lines. You can do something like
```
// Note: Non-list leaf column does not require taking into account leaf_column_offset
```
So it's clearer that the comments are separate